dask_geopandas/morton_distance.py

```diff
@@ -0,0 +1,80 @@
+import numpy as np
+import pandas as pd
+from dask_geopandas.hilbert_distance import _continuous_to_discrete_coords
+
+# Based on #http://graphics.stanford.edu/~seander/bithacks.html#InterleaveBMN
+
+
+def _morton_distance(gdf, total_bounds, p):
+    """
+    Calculate distance of geoms along Morton curve
+
+    The Morton curve is also known as Z-order https://en.wikipedia.org/wiki/Z-order
+
+    Parameters
+    ----------
+    gdf : GeoDataFrame
+    total_bounds : array_like
+        array containing xmin, ymin, xmax, ymax
+    p : int
+        precision of the Morton curve
+
+    Returns
+    -------
+    type : pandas.Series
+        Series containing distances from Morton curve
+
+    """
+
+    # Calculate bounds as numpy array
+    bounds = gdf.bounds.to_numpy()
+    # Calculate discrete coords based on total bounds and bounds
+    coords = _continuous_to_discrete_coords(total_bounds, bounds, p)
+    # Calculate distance from morton curve
+    distances = _distances_from_coordinates(coords)
+
+    return pd.Series(distances, index=gdf.index, name="morton_distance")
+
+
+def _distances_from_coordinates(coords):
+    """
+    Calculate distances from geometry mid-points along Morton curve
+
+    Parameters
+    ----------
+    coords : array_like
+        x, y coordinate pairs based on mid-points of geoms
+
+    Returns
+    -------
+    type : int
+        Integer distances from Morton curve
+    """
+
+    return np.bitwise_or(
+        _part1by1(coords[:, 0]), np.left_shift(_part1by1(coords[:, 1]), 1)
+    )
+
+
+def _part1by1(n):
+    """
+    Interleave bits by Binary Magic Numbers
+
+    Parameters
+    ----------
+    n : np.array
+        X or Y coordinates
+
+    Returns
+    -------
+    n : int
+        Interleaved bits
+    """
+
+    n &= 0x0000FFFF
```

Comment:
This line is not included in the code in http://graphics.stanford.edu/~seander/bithacks.html#InterleaveBMN (the reference you added now). Do you know what it is for? (it's eg also there in pymorton https://github.com/trevorprater/pymorton/blob/f248683c19d90e904193c58bbd03e77bc2c43768/pymorton/pymorton.py#L11)

Reply:
yes, it's also not used in the  [`Real-Time Collision Detection book, Chapter 7, 317`](http://www.r-5.org/files/books/computers/algo-list/realtime-3d/Christer_Ericson-Real-Time_Collision_Detection-EN.pdf

Updated in 27ca9e7

Reply:
Looking a little bit into it, this might actually still be needed because we are using int64 for the integer coordinates. With the current value of `p=15` that you are using for testing, the coordinates never become larger than the max value of an uint16. 

But if you try to run the tests with eg `p=20`, do they still pass?

The example code in the book you reference uses `uint32` type for the input values, and says it only uses the low 16 bits of those numbers (and then spreading those out with the bit interleaving). So if I am understanding correctly, if your coordinates are 16bit, you view then as 32bit when passing to the bit interleaving code. And that line `n &= 0x0000FFFF` might ensure that it only uses the lower 16 bits? (although I still don't fully understand why it is in there, as if you have a number that is larger than what can be represented by the 16bits (>65535), it gets truncated by `n &= 0x0000FFFF` anyway and then you also get a wrong result ..
But eg GEOS also has a Morton implementation, and includes it as well: https://github.com/libgeos/geos/blob/16610691611decd2b9545ea68ef41e8b94f859e6/src/shape/fractal/MortonCode.cpp#L95-L104

Reply:
yes, that's correct. These examples assume the input values are of type `uint32`. Currently `uint64` coords are passed to the `_part1by1` function. This comes from the `_continuous_to_discrete_coords` function. 

In #70 and #90 we discussed several follow up issues; 
1. Setting the dtype for the `_continuous_to_discrete_coords` function
2. Centralizing several functions which are used by both Hilbert and Morton distance.  

We could look into this further alongside these follow up issues and for now, simply pass the coords to the `_part1by1` function as type `unit32`? - coords.astype("uint32")

Reply:
I think this is the last component that needs updating prior to merging it. Any thoughts?

Reply:
I would maybe revert the removal of this line for now, and then list it as a follow-up to check / test more.

Otherwise, I would at least eg add a test with a higher p value as mentioned above.